x11-misc/shutter: add upload dependencies

DarthGandalf opened the pull request:
Protected by a USE-flag.

Package-Manager: Portage-2.3.89, Repoman-2.3.20
Signed-off-by: Alexey Sokolov <sokolov@google.com>
Bug: https://bugs.gentoo.org/558332

thesamesam commented:
As per the QA policy, conditional runtime dependencies are not permitted. Maybe you could just inform the user about it in pkg_postint?

https://projects.gentoo.org/qa/policy-guide/dependencies.html#pg0001

DarthGandalf commented:
That's what this ebuild does now. But it's too easy to miss the message (or forget to add it after seeing the message), and these dependencies will pollute the world set if installed manually. Whenever user tries to cleanup world set (I sometimes do that), they will probably not remember why this perl package is installed, as they don't need it directly.

I understand the rationale of that policy for usual packages, but this is pure perl, so its "compilation" time is not a problem.

Perhaps add these deps without making it optional?

DarthGandalf commented:
Another point: such message can become outdated, which happened here already: the message suggests `dev-perl/JSON-XS` while the needed package is `dev-perl/JSON-MaybeXS`, and there's no way to automatically check that.

thesamesam commented:
I understand, it's tricky in this situation given it is all pure Perl. Perhaps you could ask the QA team (or in #gentoo-proxy-maint) for some clarification?

DarthGandalf commented:
I worked around that policy. PTAL?

thesamesam commented:
Looks good -- do a revbump instead of changing the existing version so that users update.

DarthGandalf commented:
Done

DarthGandalf commented:
Diff from 0.94.3:
```diff
--- a/x11-misc/shutter/shutter-0.94.3.ebuild	2019-12-11 12:40:03.000000000 +0000
+++ b/x11-misc/shutter/shutter-0.94.3-r1.ebuild	2020-04-13 12:08:49.059832459 +0100
@@ -1,4 +1,4 @@
-# Copyright 1999-2019 Gentoo Authors
+# Copyright 1999-2020 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
@@ -12,9 +12,9 @@ SRC_URI="https://launchpad.net/shutter/0
 LICENSE="GPL-3"
 SLOT="0"
 KEYWORDS="~amd64 ~x86"
-IUSE="drawing"
+IUSE="drawing upload"
 
-RDEPEND="dev-lang/perl
+DEPEND="dev-lang/perl
 	drawing? ( dev-perl/Goo-Canvas )
 	dev-perl/libxml-perl
 	dev-perl/gnome2-canvas
@@ -37,6 +37,11 @@ RDEPEND="dev-lang/perl
 	dev-perl/X11-Protocol
 	dev-perl/XML-Simple
 	dev-perl/libwww-perl
+	upload? (
+		dev-perl/JSON-MaybeXS
+		dev-perl/Net-OAuth
+		dev-perl/Path-Class
+	)
 	virtual/imagemagick-tools[perl]"
 
 src_prepare() {
@@ -78,8 +83,6 @@ pkg_postinst() {
 	elog ""
 	elog "- media-libs/exiftool            : Writing Exif information"
 	elog "- dev-libs/libappindicator       : Status icon support for Unity"
-	elog "- dev-perl/{Net-OAuth,Path-Class}: Dropbox support"
-	elog "- dev-perl/JSON-XS               : vgy.me image hosting support"
 	elog""
 }
 

```

DarthGandalf commented:
After a discussion at #gentoo-dev-help I moved it back to RDEPEND, but also moved the rest of these optional deps there, under a single USE-flag "minimal". libappindicator is just gone, because dev-perl/ Gtk2-AppIndicator, which should be there instead, is not packaged anymore.
```diff
--- shutter-0.94.3.ebuild	2020-04-18 21:44:48.448054756 +0100
+++ shutter-0.94.3-r1.ebuild	2020-05-04 19:06:46.936593682 +0100
@@ -1,4 +1,4 @@
-# Copyright 1999-2019 Gentoo Authors
+# Copyright 1999-2020 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
@@ -12,10 +12,9 @@ SRC_URI="https://launchpad.net/shutter/0
 LICENSE="GPL-3"
 SLOT="0"
 KEYWORDS="~amd64 ~x86"
-IUSE="drawing"
+IUSE="minimal"
 
 RDEPEND="dev-lang/perl
-	drawing? ( dev-perl/Goo-Canvas )
 	dev-perl/libxml-perl
 	dev-perl/gnome2-canvas
 	dev-perl/gnome2-perl
@@ -37,12 +36,20 @@ RDEPEND="dev-lang/perl
 	dev-perl/X11-Protocol
 	dev-perl/XML-Simple
 	dev-perl/libwww-perl
-	virtual/imagemagick-tools[perl]"
+	virtual/imagemagick-tools[perl]
+	!minimal? (
+		dev-perl/Goo-Canvas
+		dev-perl/JSON-MaybeXS
+		dev-perl/Net-OAuth
+		dev-perl/Path-Class
+		media-libs/exiftool
+	)
+"
 
 src_prepare() {
 	default
 
-	use drawing || eapply "${FILESDIR}"/${PN}-0.90-goocanvas.patch
+	use minimal && eapply "${FILESDIR}"/${PN}-0.90-goocanvas.patch
 
 	#Fix tray icon because it doesn't pick the right icon using various themes
 	sed -i -e "/\$tray->set_from_icon_name/s:set_from_icon_name:set_from_file:" \
@@ -72,15 +79,6 @@ src_install() {
 pkg_postinst() {
 	xdg_icon_cache_update
 	xdg_desktop_database_update
-
-	elog ""
-	elog "These optional dependencies provide additional functions:"
-	elog ""
-	elog "- media-libs/exiftool            : Writing Exif information"
-	elog "- dev-libs/libappindicator       : Status icon support for Unity"
-	elog "- dev-perl/{Net-OAuth,Path-Class}: Dropbox support"
-	elog "- dev-perl/JSON-XS               : vgy.me image hosting support"
-	elog""
 }
 
 pkg_postrm() {

```

juippis commented on x11-misc/shutter/shutter-0.94.3-r1.ebuild:
```diff
@@ -0,0 +1,87 @@
+# Copyright 1999-2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+inherit xdg-utils desktop
```
One empty line below EAPI please, for consistency.

juippis replied:
Also please sort alphabetically since they don't collide with exporting functions.

juippis commented on x11-misc/shutter/shutter-0.94.3-r1.ebuild:
```diff
@@ -0,0 +1,87 @@
+
+DESCRIPTION="Feature-rich screenshot program"
+HOMEPAGE="https://shutter-project.org/"
+#SRC_URI="https://shutter-project.org/wp-content/uploads/releases/tars/${P}.tar.gz"
```
Can be removed.

juippis commented on x11-misc/shutter/shutter-0.94.3-r1.ebuild:
```diff
@@ -0,0 +1,87 @@
+src_prepare() {
+	default
+
+	use minimal && eapply "${FILESDIR}"/${PN}-0.90-goocanvas.patch
```
Since you need to patch off core features, maybe it'd be better to pull goo-canvas unconditionally and stop negatively patching shutter?

juippis commented on x11-misc/shutter/shutter-0.94.3-r1.ebuild:
```diff
@@ -0,0 +1,87 @@
+	dev-perl/XML-Simple
+	dev-perl/libwww-perl
+	virtual/imagemagick-tools[perl]
+	!minimal? (
```
I checked and `USE="-minimal"` pulls 100 deps, and `USE="minimal"` 90 deps. It's not *very* minimal then. I'd much rather see this turned around, into **pulling** deps with `extras` or similar USE flag.

I'd be the **happiest** if you added `upload` USE flag pulling those 3 necessary deps, and then `exiftool` added as `optfeature`. And pull `Goo-Canvas` unconditionally.

DarthGandalf replied:
I made it "drawing upload", and exiftool in optfeature. And removed the patch which was breaking the drawing feature.

juippis commented on x11-misc/shutter/shutter-0.94.3-r1.ebuild:
```diff
@@ -0,0 +1,87 @@
+	doins -r share/locale
+	insinto /usr/share/icons/hicolor
+	doins -r share/icons/hicolor/*
+	find "${D}"/usr/share/shutter/resources/system/plugins/ -type f ! -name '*.*' -exec chmod 755 {} \; \
+		|| die "failed to make plugins executables"
+	find "${D}"/usr/share/shutter/resources/system/upload_plugins/upload -type f \
+		-name "*.pm" -exec chmod 755 {} \; || die "failed to make upload plugins executables"
```
I'm not familiar with perl, but do these need to be executables? Aren't they "read through" a perl implementation?
Also perhaps make the upload chmodding related to `upload` USE flags state.

DarthGandalf commented:
<details><summary>New diff</summary>

```diff
--- shutter-0.94.3.ebuild	2020-06-07 00:14:16.472686599 +0100
+++ shutter-0.94.3-r1.ebuild	2020-06-09 21:10:49.262222333 +0100
@@ -2,20 +2,20 @@
 # Distributed under the terms of the GNU General Public License v2
 
 EAPI=7
-inherit xdg-utils desktop
+
+inherit desktop eutils xdg-utils
 
 DESCRIPTION="Feature-rich screenshot program"
-HOMEPAGE="https://shutter-project.org/"
-#SRC_URI="https://shutter-project.org/wp-content/uploads/releases/tars/${P}.tar.gz"
+HOMEPAGE="http://shutter-project.org/"
 SRC_URI="https://launchpad.net/shutter/0.9x/${PV}/+download/${P}.tar.gz"
 
 LICENSE="GPL-3"
 SLOT="0"
 KEYWORDS="~amd64 ~x86"
-IUSE="drawing"
+IUSE="drawing upload"
 
-RDEPEND="dev-lang/perl
-	drawing? ( dev-perl/Goo-Canvas )
+RDEPEND="
+	dev-lang/perl
 	dev-perl/libxml-perl
 	dev-perl/gnome2-canvas
 	dev-perl/gnome2-perl
@@ -37,35 +37,39 @@
 	dev-perl/X11-Protocol
 	dev-perl/XML-Simple
 	dev-perl/libwww-perl
-	virtual/imagemagick-tools[perl]"
-
-src_prepare() {
-	default
-
-	use drawing || eapply "${FILESDIR}"/${PN}-0.90-goocanvas.patch
-
-	#Fix tray icon because it doesn't pick the right icon using various themes
-	sed -i -e "/\$tray->set_from_icon_name/s:set_from_icon_name:set_from_file:" \
-	-e "s:shutter-panel:/usr/share/icons/hicolor/scalable/apps/&.svg:" \
-	bin/shutter || die "failed to fix trayicon"
-}
+	virtual/imagemagick-tools[perl]
+	drawing? (
+		dev-perl/Goo-Canvas
+	)
+	upload? (
+		dev-perl/JSON-MaybeXS
+		dev-perl/Net-OAuth
+		dev-perl/Path-Class
+	)
+"
 
 src_install() {
-	dobin bin/${PN}
-	insinto /usr/share/${PN}
-	doins -r share/${PN}/*
+	dobin bin/shutter
 	dodoc README
-	domenu share/applications/${PN}.desktop
+	domenu share/applications/shutter.desktop
+	doicon share/pixmaps/shutter.png
+
 	# Man page is broken. Reconstruct it.
-	gunzip share/man/man1/${PN}.1.gz || die "gunzip failed"
-	doman share/man/man1/${PN}.1
-	doicon share/pixmaps/${PN}.png
+	gunzip share/man/man1/shutter.1.gz || die "gunzip failed"
+	doman share/man/man1/shutter.1
+
+	insinto /usr/share
+	doins -r share/shutter
 	doins -r share/locale
-	insinto /usr/share/icons/hicolor
-	doins -r share/icons/hicolor/*
-	find "${D}"/usr/share/shutter/resources/system/plugins/ -type f ! -name '*.*' -exec chmod 755 {} \; \
+	doins -r share/icons
+
+	insinto /usr/share/metainfo
+	doins share/appdata/shutter.appdata.xml
+
+	find "${ED}"/usr/share/shutter/resources/system/plugins/ -type f ! -name '*.*' -exec chmod 755 {} \; \
 		|| die "failed to make plugins executables"
-	find "${D}"/usr/share/shutter/resources/system/upload_plugins/upload -type f \
+	# shutter executes perl scripts as standalone scripts, and after that "require"s them.
+	find "${ED}"/usr/share/shutter/resources/system/upload_plugins/upload -type f \
 		-name "*.pm" -exec chmod 755 {} \; || die "failed to make upload plugins executables"
 }
 
@@ -73,14 +77,7 @@
 	xdg_icon_cache_update
 	xdg_desktop_database_update
 
-	elog ""
-	elog "These optional dependencies provide additional functions:"
-	elog ""
-	elog "- media-libs/exiftool            : Writing Exif information"
-	elog "- dev-libs/libappindicator       : Status icon support for Unity"
-	elog "- dev-perl/{Net-OAuth,Path-Class}: Dropbox support"
-	elog "- dev-perl/JSON-XS               : vgy.me image hosting support"
-	elog""
+	optfeature "Writing Exif information" media-libs/exiftool
 }
 
 pkg_postrm() {

```
</details>

juippis commented on x11-misc/shutter/shutter-0.94.3-r1.ebuild:
```diff
@@ -0,0 +1,86 @@
+	dev-perl/XML-Simple
+	dev-perl/libwww-perl
+	virtual/imagemagick-tools[perl]
+	drawing? (
+		dev-perl/Goo-Canvas
+	)
```
Just add it as optfeature...
